### PR TITLE
[api-minor] Add a parameter to `PDFPageProxy_getTextContent` that enables replacing of all whitespace with standard spaces in the textLayer (issue 6612)

### LINKED ISSUE #6612: Wrong Text Layer rendering for some PDFs
Here is a sample PDF: https://github.com/mpankov/pdfjs_selection_issue/raw/master/test_document.pdf
Text Layer words position doesn't correspond to canvas.
Search highlight is outside of text:
wrong_selection_position

As I can see PDF.js renders "tab" symbols in Text Layer (instead of spaces) and that makes text longer...

### PR DESCRIPTION
**Pull request by Snuffleupagus**
This patch goes a bit further than issue #6612 requires, and replaces all kinds of whitespace with standard spaces.

When testing this locally, it actually seemed to slightly improve two existing test-cases (`tracemonkey-text` and `taro-text`).

Fixes #6612.

_Note:_ The WIP in the title is because I've not yet added a reduced test-case, and because I'm not sure if we want to replace every kind of whitespace in this way.

### CODE REVIEW

File: src/display/text_layer.js [hunk 1]
```diff
@@ -75,7 +81,7 @@ var renderTextLayer = (function renderTextLayerClosure() {
     textDiv.style.fontSize = fontHeight + 'px';
     textDiv.style.fontFamily = style.fontFamily;
 
-    textDiv.textContent = geom.str;
+    textDiv.textContent = replaceWhitespaceWithStandardSpace(geom.str);
```
Comment (yurydelendik) [inline src/display/text_layer.js]: Could you add a comment about why we are doing that? There is actually a second reason `measureText` will replace all white spaces to SPACE anyway.

**Comment by yurydelendik** [comment]
good with comment adressed

/botio test

**Comment by Snuffleupagus** [comment]
I just realized that in its current form, this patch suffers from a quite annoying issue.
Even though the textLayer is rendered correctly, when you try to search, the textLayer reverts to the previous broken alignment.
Even though the initial textLayer `<div>` has replaced the tabs, both `TextLayerBuilder` and `PDFFindController` have a reference to the result of [`getTextContent`](https://github.com/mozilla/pdf.js/blob/master/src/display/api.js#L887), which obviously contains the "original" textLayer content without the whitespace conversion.

Replacing whitespaces with a regexp at every point where the viewer code is doing `getTextContent` seems quite cumbersome and ineffective, so I'm wondering if we could perhaps extend the API by adding an parameter for [`getTextContent`](https://github.com/mozilla/pdf.js/blob/master/src/display/api.js#L887) that will return the text with the whitespace conversion done inside of e.g. https://github.com/mozilla/pdf.js/blob/master/src/core/evaluator.js#L911 instead?

@yurydelendik How do you think that we should proceed here?

**Comment by yurydelendik** [comment]
Okay, somewhere at [runBidiTransform](https://github.com/mozilla/pdf.js/blob/master/src/core/evaluator.js#L1033) there will be a good place for it.

Can you also check what is faster: replace with regex or split.join with regex? Also I think it will worth to check if all characters belong to specific range before doing the replacement in regular for-loop, e.g. with charCode >= 32 and < 127.

**Comment by Snuffleupagus** [comment]
I've pushed a new version of the patch; please refer to https://gist.github.com/Snuffleupagus/15309d9709690141e702 for a quick comparison between different replacement techniques.

**Comment by yurydelendik** [comment]
Strange that checking against good range of characters did not work (check also http://jsperf.com/replace-spaces)

**Comment by Snuffleupagus** [comment]
> Strange that checking against good range of characters did not work (check also http://jsperf.com/replace-spaces)

Yeah, I've now tested this in the actual code, and it seems that it is *slightly* faster to check the character range (even if the difference is small). I'll update the patch, thanks for the test-case!

File: src/core/evaluator.js [hunk 2]
```diff
@@ -1030,8 +1033,22 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
...
+        if (normalizeWhitespace) {
+          // Replace all whitespaces with standard spaces (0x20), to avoid
+          // alignment issues between the textLayer and the canvas if the text
+          // contains e.g. tabs (fixes issue6612.pdf).
+          var i = 0, ii = bidiResult.str.length, charCode;
+          while (i < ii && (charCode = bidiResult.str.charCodeAt(i)) >= 0x20 &&
+                 charCode <= 0x7F) {
+            i++;
+          }
+          shouldReplaceWhitespace = i < ii;
```
Comment (yurydelendik) [inline src/core/evaluator.js]: Create a function that will handle the whitespace replacement? Please notice to be compiled function must be short and invoke lots of times. The inlining may not produce compiled result hence make loop with charCodeAt show.

Comment (Snuffleupagus) [inline src/core/evaluator.js]: Good point, I've pushed a new version that addresses this.

**Comment by yurydelendik** [comment]
Looks good.

/botio test

**Comment by yurydelendik** [comment]
(Hmmm, I wonder if our makeref works.)

/botio makeref

Thank you for the patch.

**Comment by yurydelendik** [comment]
Hmmm, actually it might be a regression, investigating

**Comment by yurydelendik** [comment]
Looks good locally, monitoring that as intermittent.

**Pull request merged**